### PR TITLE
fix: vesting logic

### PR DESCRIPTION
**Pull request by djm07073**
- fix bug of vested position reward remain 
- fix average tvl because of overlapping snapshot and overflow by multiple snapshots

## Summary by CodeRabbit

- **New Features**
  - Enhanced vesting functions to include penalty rewards.
  - Added two new public functions to check if vesting positions are finalized.
  - Introduced new functions for batch claiming and creating user vestings.
  - Added new events to signal user vesting record creation.

- **Improvements**
  - Updated multiple function signatures and return types for consistency.
  - Changed terminology from `proportion` to `minimum_score_ratio` for clarity.
  - Updated terms related to epochs, reflecting a shift to cycle-based terminology.

- **Refactor**
  - Consistent renaming of fields and constants across modules for clarity and cohesion.

### CODE REVIEW

**Comment by djm07073** [comment]
i'll add some few test cases of vesting finalized

File: precompile/modules/initia_stdlib/sources/vip/tvl_manager.move [hunk 1]
```diff
@@ -34,162 +34,187 @@ module initia_std::vip_tvl_manager {
...
+        // update the average tvl of the bridge at the stage
+        let average_tvl = table::borrow_mut(
             average_tvl_table,
-            table_key::encode_u64(count),
-            snapshot
+            table_key::encode_u64(bridge_id)
         );
+        // new average tvl = (snapshot_count * average_tvl + balance) / (snapshot_count + 1)
+        let new_average_tvl = decimal256::mul_u64(
+            &decimal256::from_ratio_u64(*average_tvl,(snapshot_count + 1)),
```
Comment (beer-1) [inline precompile/modules/initia_stdlib/sources/vip/tvl_manager.move:127]: seems don't need to divide it?

File: precompile/modules/initia_stdlib/sources/vip/vesting.move [hunk 2]
```diff
@@ -323,451 +287,625 @@ module initia_std::vip_vesting {
...
+            let value = vector::borrow_mut(user_vestings, idx);
+            let vest_amount = if (claim_info.l2_score >= value.minimum_score) {value.vest_max_amount} else {
+                (
+                    (value.vest_max_amount as u128) * (claim_info.l2_score as u128) / (
+                        value.minimum_score as u128
+                    ) as u64
+                )
+            };
+            vested_reward = vested_reward + vest_amount;
+            penalty_reward = penalty_reward + value.vest_max_amount - vest_amount;
```
Comment (beer-1) [inline precompile/modules/initia_stdlib/sources/vip/vesting.move:424]: This penalty_reward seems cumulative and this cumulative value used to build penalty_map, is it intended? just for checking.

File: precompile/modules/initia_stdlib/sources/vip/vesting.move [hunk 3]
```diff
@@ -323,451 +287,625 @@ module initia_std::vip_vesting {
...
+                    vector::push_back(penalty_keys, value.start_stage);
+                };
+
+                let penalty_amount = simple_map::borrow_mut(penalty_map, &value.start_stage);
+                *penalty_amount = *penalty_amount + penalty_reward;
+            };
 
-            // move vesting if end stage is over or the left reward is empty
-            if (stage > value.end_stage || value.remaining_reward == 0) {
+            if (claim_info.start_stage >= value.end_stage) {
```
Comment (beer-1) [inline precompile/modules/initia_stdlib/sources/vip/vesting.move:437]: I think this check should be done right after we borrow vesting value and do emit event and `continue` if the stage is out of range and change this line to just checking `claim_info.start_stage == value.end_stage`.

File: precompile/modules/initia_stdlib/sources/vip/vesting.move [hunk 4]
```diff
@@ -323,451 +287,625 @@ module initia_std::vip_vesting {
...
+                        table_key::encode_u64(start_stage)
                     ),
                     error::unavailable(EVESTING_NOT_CLAIMED)
                 );
 
-                let vesting = table::remove(
-                    &mut vesting_store.vestings,
-                    table_key::encode_u64(*stage)
-                );
+                if (table::contains(
```
Comment (beer-1) [inline precompile/modules/initia_stdlib/sources/vip/vesting.move:470]: seems unnecessary, but there was some edge case maybe?

File: precompile/modules/initia_stdlib/sources/vip/vesting.move [hunk 5]
```diff
@@ -323,451 +287,625 @@ module initia_std::vip_vesting {
...
-            if (stage > value.end_stage || value.remaining_reward == 0) {
+        let finalized_vestings_idx: vector<u64> = vector[]; // vector index
+        let idx = 0;
+        let len = vector::length(operator_vestings);
+        while (idx < len) {
+            let value = vector::borrow_mut(operator_vestings, idx);
+            vested_reward = vested_reward + value.vest_max_amount;
+            value.remaining_reward = value.remaining_reward - value.vest_max_amount;
+
+            if (claim_info.start_stage >= value.end_stage) {
```
Comment (beer-1) [inline precompile/modules/initia_stdlib/sources/vip/vesting.move:508]: also good to check the vesting is out of range first.

**Comment by djm07073** [comment]
this pr is closed, however vip task in the initiation branch has been completed, and this PR will be reopened.

**Pull request closed without merging**